nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-reporting-task/src/main/java/org/apache/nifi/reporting/riemann/RiemannReportingTask.java

```diff
@@ -0,0 +1,244 @@
+        properties.add(TRANSPORT_PROTOCOL);
+        properties.add(SERVICE_PREFIX);
+        properties.add(HOSTNAME);
+        properties.add(TAGS);
+        properties.add(SEND_JVM_METRICS);
+        properties.add(SEND_NIFI_METRICS);
+        properties.add(SEND_PROCESSOR_METRICS);
+        properties.add(SEND_BULLETIN_MESSAGES);
+        properties.add(MIN_BULLETIN_LEVEL);
+        properties.add(WRITE_TIMEOUT);
+        return Collections.unmodifiableList(properties);
+    }
+
+    @Override
+    protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String propertyDescriptorName) {
+        return new PropertyDescriptor.Builder().name(propertyDescriptorName).required(false).addValidator(StandardValidators.NON_EMPTY_VALIDATOR).expressionLanguageSupported(false).dynamic(true)
+                .build();
+    }
+
+    @OnStopped
+    public final void cleanUpClient() {
+        if (riemannClient != null) {
+            this.riemannClient.close();
+        }
+        this.riemannClient = null;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ConfigurationContext context) throws ProcessException {
+        if (riemannClient == null || !riemannClient.isConnected()) {
+            transport = Transport.valueOf(context.getProperty(TRANSPORT_PROTOCOL).getValue());
+            String host = context.getProperty(RIEMANN_HOST).evaluateAttributeExpressions().getValue();
+            int port = context.getProperty(RIEMANN_PORT).asInteger();
+            RiemannClient client = null;
+            try {
+                switch (transport) {
+                case TCP:
+                    client = RiemannClient.tcp(host, port);
+                    break;
+                case UDP:
+                    client = RiemannClient.udp(host, port);
+                    break;
+                }
+                client.connect();
+                riemannClient = client;
+            } catch (IOException e) {
+                if (client != null) {
+                    client.close();
+                }
+                throw new ProcessException(String.format("Unable to connect to Riemann [%s:%d] (%s)\n%s", host, port, transport, e.getMessage()));
+            }
+        }
+        metricsService = new MetricsService(context.getProperty(SERVICE_PREFIX).getValue(), context.getProperty(HOSTNAME).evaluateAttributeExpressions().getValue(),
+                context.getProperty(TAGS).getValue().split(","));
+
+        // Add dynamic properties as attributes to messages
+        Map<String, String> additionalProperties = Maps.newHashMapWithExpectedSize(context.getProperties().size());
+        for (PropertyDescriptor propertyDescriptor : context.getProperties().keySet()) {
+            if (propertyDescriptor.isDynamic()) {
+                additionalProperties.put(propertyDescriptor.getName(), context.getProperty(propertyDescriptor).getValue());
+            }
+        }
+        metricsService.setAdditionalAttributes(additionalProperties);
```

Comment:
consider just making setting the additional properties part of the constructor of MetricsService.  Doing so it seems would make that class thread safe as well.